src/bounded_queue.ml

```diff
@@ -0,0 +1,113 @@
+(*
+  structure node t value: data type, next: pointer to node t
+  structure queue t Head: pointer to node t, Tail: pointer to node t, H lock: lock type, T lock: lock type
+*)
+
+type 'a node = Nil | Next of 'a * 'a node ref
+
+type 'a t = {
+  mutable head : 'a node ref;
+  mutable tail : 'a node ref;
+  head_lock : Mutex.t;
+  tail_lock : Mutex.t;
+  not_empty : Condition.t;
+  not_full : Condition.t;
+  size : int Atomic.t;
+  capacity : int;
+}
+
+(* create and initialize the 2 lock queue *)
+let create () =
+  let dummy = ref Nil in
+  {
+    head = dummy;
+    tail = dummy;
+    head_lock = Mutex.create ();
+    tail_lock = Mutex.create ();
+    not_empty = Condition.create ();
+    not_full = Condition.create ();
+    size = Atomic.make 0;
+    capacity = 1_000_000;
```

Comment:
Something like `Int.max_int` could be used as a default value to mean that the queue has unlimited capacity (memory will run out before).  It is then also possible to make it so that with the unlimited capacity value the capacity checks are skipped.  With a scalable length tracking this could make the bounded versions (nearly) as fast as an unbounde version.

Reply:
what about the empty case, that will still have to be checked even for this default value

Reply:
Yes, with "capacity checks" I was referring to needing to check that the capacity of the queue is not exhausted (i.e. not_full).